### PR TITLE
(take 2) SR-0140: Bridge Optionals to nonnull ObjC objects by bridging their payload, or using a sentinel.

### PR DESCRIPTION
**Pull request by jckarter**
id-as-Any lets you pass Optional to an ObjC API that takes `nonnull id`, and also lets you bridge containers of `Optional` to `NSArray` etc. When this occurs, we can unwrap the value and bridge it so that inhabited optionals still pass into ObjC in the expected way, but we need something to represent `none` other than the `nil` pointer. Cocoa provides `NSNull` as the canonical "null for containers" object, which is the least bad of many possible answers. If we happen to have the rare nested optional `T??`, there is no precedented analog for these in Cocoa, so just generate a unique sentinel object to preserve the `nil`-ness depth so we at least don't lose information round-tripping across the ObjC-Swift bridge.

Making Optional conform to _ObjectiveCBridgeable is more or less enough to make this all work, though there are a few additional edge case things that need to be fixed up. We don't want to accept `AnyObject??` as an @objc-compatible type, so special-case Optional in `getForeignRepresentable`.

Implements SR-0140 (rdar://problem/27905315).

### CODE REVIEW

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@jrose-apple Do you mind reviewing this for inclusion in 3.0.1?

File: stdlib/public/core/Optional.swift [hunk 1]
```diff
@@ -487,3 +487,78 @@ extension Optional {
...
 }
+
+//===----------------------------------------------------------------------===//
+// Bridging
+//===----------------------------------------------------------------------===//
+
+#if _runtime(_ObjC)
+extension Optional : _ObjectiveCBridgeable {
+  // The object that represents `none` for an Optional of this type.
+  internal static var _nilSentinel : AnyObject {
```
Comment (jrose-apple) [inline stdlib/public/core/Optional.swift]: Is this worth caching, for SIL optimization purposes?

```swift
internal static let _nilSentinel = _swift_Foundation_getOptionalNilSentinelObject()
```

File: stdlib/public/core/Optional.swift [hunk 2]
```diff
@@ -487,3 +487,78 @@ extension Optional {
...
+    // Bridge a wrapped value by unwrapping.
+    if let value = self {
+      return _bridgeAnythingToObjectiveC(value)
+    }
+    // Bridge nil using a sentinel.
+    return type(of: self)._nilSentinel
+  }
+
+  public static func _forceBridgeFromObjectiveC(_ source: AnyObject,
+                                         result: inout Optional<Wrapped>?) {
```
Comment (jrose-apple) [inline stdlib/public/core/Optional.swift]: Nitpick: unconventional alignment here. Standard library style would be to treat the parameters as a block bound with parens.

File: stdlib/public/core/Optional.swift [hunk 3]
```diff
@@ -487,3 +487,78 @@ extension Optional {
...
+    // Map the nil sentinel back to .none.
+    // NB that the signature of _forceBridgeFromObjectiveC adds another level
+    // of optionality, so we need to wrap the immediate result of the conversion
+    // in `.some`.
+    if source === _nilSentinel {
```
Comment (jrose-apple) [inline stdlib/public/core/Optional.swift]: Are all NSNulls guaranteed to be identical?

Comment (jrose-apple) [inline stdlib/public/core/Optional.swift]: It is documented as a singleton, but I would double-check with someone from Foundation. @phausler?

File: stdlib/public/runtime/Casting.cpp [hunk 4]
```diff
@@ -2513,6 +2513,18 @@ bool swift::swift_dynamicCast(OpaqueValue *dest,
     // unwrapping the target. This handles an optional source wrapped within an
     // existential that Optional conforms to (Any).
     if (auto srcExistentialType = dyn_cast<ExistentialTypeMetadata>(srcType)) {
+#if SWIFT_OBJC_INTEROP
+      // If coming from AnyObject, we may want to bridge.
```
Comment (jrose-apple) [inline stdlib/public/runtime/Casting.cpp]: Why is this specific to AnyObject?

File: stdlib/public/stubs/OptionalBridgingHelper.mm [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+}
+@end
+
+@implementation _SwiftNull : NSObject
+
+- (NSString*)description {
+  return [NSString stringWithFormat:@"<_SwiftNull %u>", self->depth];
+}
+
+- (void)dealloc {
```
Comment (jrose-apple) [inline stdlib/public/stubs/OptionalBridgingHelper.mm]: Nitpick: This is conventionally omitted when empty (and may allow run-time optimizations).

File: test/stdlib/OptionalBridge.swift [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+  case (.some(let xx), .some(let yy)):
+    expectEqual(xx, yy)
+  case (.none, .none):
+    return
+  default:
+    expectUnreachable("\(T.self)??? values don't match: \(x) vs. \(y)")
+  }
+}
+
+tests.test("wrapped value") {
```
Comment (jrose-apple) [inline test/stdlib/OptionalBridge.swift]: Any tests for optionals of non-bridged types? For optionals of reference types?

File: stdlib/public/stubs/OptionalBridgingHelper.mm [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include <Foundation/Foundation.h>
```
Comment (jrose-apple) [inline stdlib/public/stubs/OptionalBridgingHelper.mm]: Nitpick: I think this counts as a system include, and so it should go at the bottom. Also, `#import` would be more appropriate. (ObjC headers typically don't have include guards.)

**Comment by jrose-apple** [comment]
Some comments made, but mostly seems fine to me…except that someone else should take a look at the lib/SIL/ changes.

**Comment by atrick** [comment]
lib/SIL looks superficially ok to me. Question: Does SILGenBridging implicitly assume that the return value can't be reabstracted? If so, a comment would be helpful.

File: stdlib/public/core/Optional.swift (hunk 1 above)
Comment (jckarter) [inline stdlib/public/core/Optional.swift]: Unfortunately, we don't support stored static properties on generic types. We have to rely on the runtime's caching.

File: stdlib/public/core/Optional.swift (hunk 3 above)
Comment (jckarter) [inline stdlib/public/core/Optional.swift]: There's example ObjC code that uses `x == [NSNull null]` too.

File: stdlib/public/runtime/Casting.cpp (hunk 4 above)
Comment (jckarter) [inline stdlib/public/runtime/Casting.cpp]: I suppose we could also support it from `NSNull`. `AnyObject` is the only Swift-level formal type that'd make sense to cast back to `Optional` otherwise.

File: stdlib/public/stubs/OptionalBridgingHelper.mm (hunk 5 above)
Comment (jckarter) [inline stdlib/public/stubs/OptionalBridgingHelper.mm]: Oops, I had put this in so I could put breakpoints there, but forgot to remove it.

**Comment by jckarter** [comment]
@atrick Yeah, it looks like SILGenBridging doesn't handle reabstraction (or address-only types, or indirect arguments, which @shajrawi is fixing). I can add that as a comment.

File: stdlib/public/runtime/Casting.cpp (hunk 4 above)
Comment (jrose-apple) [inline stdlib/public/runtime/Casting.cpp]: Oh, I somehow missed that this was only when casting to Optional. Thanks for clarifying.

File: stdlib/public/core/Optional.swift (hunk 3 above)
Comment (phausler) [inline stdlib/public/core/Optional.swift]: NSNull in Darwin is a singleton

File: stdlib/public/stubs/OptionalBridgingHelper.mm [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+#include "swift/Runtime/Metadata.h"
+#include "swift/Runtime/Mutex.h"
+#include "swift/Runtime/ObjCBridge.h"
+#include <vector>
+
+using namespace swift;
+
+/// Class of sentinel objects used to represent the `nil` value of nested
+/// optionals.
+@interface _SwiftNull : NSObject {
```
Comment (phausler) [inline stdlib/public/stubs/OptionalBridgingHelper.mm]: this should'd probably be `__attribute__((visibility("hidden")))`

File: stdlib/public/stubs/OptionalBridgingHelper.mm [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+@public
+  unsigned depth;
+}
+@end
+
+@implementation _SwiftNull : NSObject
+
+- (NSString*)description {
+  return [NSString stringWithFormat:@"<_SwiftNull %u>", self->depth];
```
Comment (phausler) [inline stdlib/public/stubs/OptionalBridgingHelper.mm]: Usually NSObjects have the format more like:

`[NSString stringWithFormat:@"<%@ %p depth = %u>", [self class], self, self->depth]`

That way if there is ever a problem you have the pointer which can be used in a debugger if need be.

File: test/stdlib/OptionalBridge.swift [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+  let tripleNullBridged = tripleNull as AnyObject
+
+  expectTrue(doubleNullBridged !== tripleNullBridged)
+
+  let tripleNullCastBack = cast(tripleNullBridged, to: Optional<String??>.self)
+  expectEqual(tripleNullCastBack, tripleNull)
+}
+
+tests.test("collection of Optional") {
+  let holeyArray: [LifetimeTracked?] = [LifetimeTracked(0), nil, LifetimeTracked(1)]
```
Comment (phausler) [inline test/stdlib/OptionalBridge.swift]: Does the reverse work? `someNSArray as [LifetimeTracked?]`

This pattern if ever used will never really be portable to swift-corelibs-foundation which is a bit unfortunate.

File: stdlib/public/stubs/OptionalBridgingHelper.mm [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+  std::vector<id> Cache;
+  StaticReadWriteLock Lock;
+};
+
+static Lazy<SwiftNullSentinelCache> Sentinels;
+
+static id getSentinelForDepth(unsigned depth) {
+  // For unnested optionals, use NSNull.
+  if (depth == 1)
+    return SWIFT_LAZY_CONSTANT([NSNull null]);
```
Comment (phausler) [inline stdlib/public/stubs/OptionalBridgingHelper.mm]: What does this macro do?

File: stdlib/public/stubs/OptionalBridgingHelper.mm [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+  {
+    StaticScopedWriteLock lock(theSentinels.Lock);
+    if (depthIndex >= theSentinels.Cache.size())
+      theSentinels.Cache.resize(depthIndex + 1);
+    auto &cached = theSentinels.Cache[depthIndex];
+    // Make sure another writer didn't sneak in.
+    if (!cached) {
+      auto sentinel = [[_SwiftNull alloc] init];
+      sentinel->depth = depth;
+      cached = sentinel;
```
Comment (phausler) [inline stdlib/public/stubs/OptionalBridgingHelper.mm]: Am I not reading this correctly? On cache miss you are not storing a value into the cache?

Comment (jckarter) [inline stdlib/public/stubs/OptionalBridgingHelper.mm]: `cached` is a reference, so this will store the value into the cache.

File: test/stdlib/OptionalBridge.swift (hunk 10 above)
Comment (jckarter) [inline test/stdlib/OptionalBridge.swift]: That's true of all bridging conversions, but ideally the bridging is never necessary to begin with in corelibs.

File: stdlib/public/stubs/OptionalBridgingHelper.mm (hunk 11 above)
Comment (phausler) [inline stdlib/public/stubs/OptionalBridgingHelper.mm]: I am fairly certain we have a full obligation from Foundation to maintain the return value as implemented:

```objc
+ (NSNull *)null {
    return (id)kCFNull;
}
```

So the dispatch_once here is not really needed at all.

Comment (jckarter) [inline stdlib/public/stubs/OptionalBridgingHelper.mm]: Good idea, I'll just use `kCFNull` directly.

File: stdlib/public/stubs/OptionalBridgingHelper.mm (hunk 8 above)
Comment (jckarter) [inline stdlib/public/stubs/OptionalBridgingHelper.mm]: Good idea. We aren't making any of the runtime-private classes hidden yet. I think we should do that as a separate patch, though, just to keep the risk of unintentional breakage minimal for this as a 3.0.1 patch.

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@jrose-apple @phausler @atrick Thanks for the review comments! I've revised the patch in response, if you all don't mind taking another look.

**Comment by jckarter** [comment]
@swift-ci Please test

File: lib/AST/Builtins.cpp [hunk 13]
```diff
@@ -754,6 +754,7 @@ static ValueDecl *getIsOptionalOperation(ASTContext &Context, Identifier Id) {
   return builder.build(Id);
 }
 
+
```
Comment (jrose-apple) [inline lib/AST/Builtins.cpp]: Nitpick: this is the only change in this file now.

**Comment by jrose-apple** [comment]
LGTM

File: lib/AST/Builtins.cpp (hunk 13 above)
Comment (jckarter) [inline lib/AST/Builtins.cpp]: Thanks, fixed.

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test
